File: ruby/medium/kth_smallest_element_in_a_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    counter_map = {}
    count_nodes(root, counter_map)
    return find_kth(root, 0, k, counter_map).val
end

def find_kth(root, i, k, counter_map)
    return nil unless root
    count = counter_map[root]
    n = (i + count[:left] + 1)
    if n == k
        return root
    elsif k < n
        return find_kth(root.left, i, k, counter_map)
    else
        return find_kth(root.right, n, k, counter_map)
    end
end

def count_nodes(root, counter_map)
    return 0 unless root
    left = count_nodes(root.left, counter_map)
    right = count_nodes(root.right, counter_map)
    counter_map[root] = { left: left, right: right }
    return 1 + left + right
end

File: javascript/medium/kthSmallestElementInABST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    const map = new Map();
    countNodes(root, map);
    return findKth(root, k, map).val;
};

var findKth = function(root, k, map) {
    if(root === null) return null;
    const count = map.get(root);
    const n = count[0] + 1;
    if(n === k) {
        return root;
    } else if(k < n) {
        return findKth(root.left, k, map);
    } else {
        return findKth(root.right, (k-n), map);
    }
};

var countNodes = function(root, map) {
    if(root === null) return 0;
    const left = countNodes(root.left, map);
    const right = countNodes(root.right, map);
    map.set(root, [left, right]);
    return 1 + left + right;
};
